cmd/kar-controllers/app/server.go

```diff
@@ -76,14 +84,69 @@ func Run(opt *options.ServerOption) error {
 }
 
 // Starts the health probe listener
-func listenHealthProbe(opt *options.ServerOption) error {
-	handler := http.NewServeMux()
-	handler.Handle("/healthz", &health.Handler{})
-	err := http.ListenAndServe(opt.HealthProbeListenAddr, handler)
-	if err != nil {
-		return err
-	}
-
-	return nil
+func startHealthAndMetricsServers(opt *options.ServerOption) error {
+    // Start the metrics server
+    RecordMetrics()
+
+    metricsHandler := http.NewServeMux()
+    metricsHandler.Handle("/metrics", promhttp.Handler())
+    healthHandler := http.NewServeMux()
+    healthHandler.Handle("/healthz", &health.Handler{})
+
+    metricsServer := &http.Server{
+        Addr:    opt.MetricsListenAddr,
+        Handler: metricsHandler,
+    }
+
+    healthServer := &http.Server{
+        Addr:    opt.HealthProbeListenAddr,
+        Handler: healthHandler,
+    }
+
+    // make a channel for errors for each server
+    metricsServerErrChan := make(chan error)
+    healthServerErrChan := make(chan error)
+
+    // start servers in their own goroutines
+    go func() {
+        err := metricsServer.ListenAndServe()
+        if err != nil && err != http.ErrServerClosed {
+            metricsServerErrChan <- err
+        }
+    }()
+
+    go func() {
+        err := healthServer.ListenAndServe()
+        if err != nil && err != http.ErrServerClosed {
+            healthServerErrChan <- err
+        }
+    }()
+
+    // make a channel to listen for OS shutdown signal
+    stop := make(chan os.Signal, 1)
```

Comment:
That should be handled in the main.go file. I'd suggest creating a root context with `k8s.io/apiserver/pkg/server.SetupSignalContext()` there, and passing it here.

We also need to make sure the metrics and health servers are the latest to be shutdown as the process terminate.